### PR TITLE
bake: add wildcard to fs entitlements to allow any paths

### PR DESCRIPTION
**Pull request by crazy-max**
follow-up https://github.com/docker/buildx/pull/2796#issuecomment-2491867068

### CODE REVIEW

**Comment by crazy-max** [comment]
Removed long path name handling in our tests https://github.com/docker/buildx/pull/2812/commits/619fa8ad4c487c0793cdf87c92c4db19ac20b282 but it fails: https://github.com/docker/buildx/actions/runs/11972131393/job/33378391055#step:6:1330

```
 === Failed
=== FAIL: bake TestValidateEntitlements/ExportLocal (0.00s)
    entitlements_test.go:352: 
        	Error Trace:	D:/a/buildx/buildx/bake/entitlements_test.go:352
        	Error:      	Not equal: 
        	            	expected: bake.EntitlementConf{NetworkHost:false, SecurityInsecure:false, FSRead:[]string{"D:\\a\\buildx\\buildx\\bake"}, FSWrite:[]string{"C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\001", "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\002"}, ImagePush:[]string(nil), ImageLoad:[]string(nil), SSH:false}
        	            	actual  : bake.EntitlementConf{NetworkHost:false, SecurityInsecure:false, FSRead:[]string{"D:\\a\\buildx\\buildx\\bake"}, FSWrite:[]string{"C:\\Users\\runneradmin\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\001", "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\002"}, ImagePush:[]string(nil), ImageLoad:[]string(nil), SSH:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -7,4 +7,4 @@
        	            	  FSWrite: ([]string) (len=2) {
        	            	-  (string) (len=75) "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\001",
        	            	-  (string) (len=75) "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\002"
        	            	+  (string) (len=78) "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\001",
        	            	+  (string) (len=78) "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\TestValidateEntitlements4101399922\\002"
        	            	  },
        	Test:       	TestValidateEntitlements/ExportLocal
    --- FAIL: TestValidateEntitlements/ExportLocal (0.00s)

=== FAIL: bake TestValidateEntitlements/SecretFromEscapeLink (0.00s)
    entitlements_test.go:352: 
        	Error Trace:	D:/a/buildx/buildx/bake/entitlements_test.go:352
        	Error:      	Not equal: 
        	            	expected: bake.EntitlementConf{NetworkHost:false, SecurityInsecure:false, FSRead:[]string{"C:\\Users\\RUNNER~1\\AppData\\Local\\Temp"}, FSWrite:[]string(nil), ImagePush:[]string(nil), ImageLoad:[]string(nil), SSH:false}
        	            	actual  : bake.EntitlementConf{NetworkHost:false, SecurityInsecure:false, FSRead:[]string{"C:\\Users\\runneradmin\\AppData\\Local\\Temp"}, FSWrite:[]string(nil), ImagePush:[]string(nil), ImageLoad:[]string(nil), SSH:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	  FSRead: ([]string) (len=1) {
        	            	-  (string) (len=36) "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp"
        	            	+  (string) (len=39) "C:\\Users\\runneradmin\\AppData\\Local\\Temp"
        	            	  },
        	Test:       	TestValidateEntitlements/SecretFromEscapeLink
    --- FAIL: TestValidateEntitlements/SecretFromEscapeLink (0.00s)

=== FAIL: bake TestValidateEntitlements (1.07s)
```

It makes sense as the expected paths don't match anymore like https://github.com/docker/buildx/blob/17eff25fe591ef7ecf0a7046cb55f847db42f575/bake/entitlements_test.go#L279

So as `filepath.EvalSymlinks` evaluates itself to a long path name in entitlements implementation we don't have the choice to handle it in our tests as well.

Here is a repro of the behavior: https://github.com/crazy-max/go-longpath-win/blob/435f1ae22f324f213b7c7b49be7165aa9eafbd3d/main.go

```go
tmpDir := os.TempDir()
wd, _ := os.Getwd()
fmt.Printf("TempDir: %s\n", tmpDir)
fmt.Printf("WorkingDir: %s\n", wd)

tmpDirSym, _ := filepath.EvalSymlinks(tmpDir)
wdSym, _ := filepath.EvalSymlinks(wd)
fmt.Printf("TempDirEvalSymlinks: %s\n", tmpDirSym)
fmt.Printf("WorkingDirEvalSymlinks: %s\n", wdSym)
```

Output on GHA: https://github.com/crazy-max/go-longpath-win/actions/runs/11972415653/job/33379286983#step:4:7

```
TempDir: C:\Users\RUNNER~1\AppData\Local\Temp
WorkingDir: D:\a\go-longpath-win\go-longpath-win
TempDirEvalSymlinks: C:\Users\runneradmin\AppData\Local\Temp
WorkingDirEvalSymlinks: D:\a\go-longpath-win\go-longpath-win
```

File: bake/entitlements_test.go [hunk 1]
```diff
@@ -343,6 +343,31 @@ func TestValidateEntitlements(t *testing.T) {
...
+					// on windows root (/) is only allowed if it is the same volume as wd
+					if filepath.VolumeName(wd) == filepath.VolumeName(escapeLink) {
+						return nil
+					}
+					// if not, then escapeLink is not allowed
+					p, err := evaluateToExistingPath(escapeLink)
+					require.NoError(t, err)
+					return []string{p}
+				}(),
+			},
```
Comment (crazy-max) [inline bake/entitlements_test.go:357]: Revert the change to allow any volumes for `/` on windows and handle that case in our tests instead.

File: bake/entitlements_windows.go [hunk 2]
```diff
@@ -10,7 +10,7 @@ import (
 func evaluatePaths(in []string) ([]string, error) {
 	out := make([]string, 0, len(in))
 	for _, p := range in {
-		if p == "/" {
+		if p == "*" {
```
Comment (crazy-max) [inline bake/entitlements_windows.go:13]: Introduce wildcard pattern to allow any volumes on windows. On unix it would be `/`.

Comment (tonistiigi) [reply bake/entitlements_windows.go:13]: I think we should just handle `*` as a special value that acts as a valid parent for any path, rather than try to convert `*` into a list of all possible volumes on windows.

File: bake/entitlements.go [hunk 3]
```diff
@@ -335,9 +335,11 @@ func findMissingPaths(set []string, paths map[string]struct{}) ([]string, error)
 		return nil, err
 	}
 
-	set, err = evaluatePaths(set)
+	set, allowAny, err := evaluatePaths(set)
```
Comment (tonistiigi) [inline bake/entitlements.go:338]: nit: this could be moved before the `evaluateToExistingPaths`

**Review by tonistiigi — APPROVED** [review]
